### PR TITLE
[FLINK-14280] Introduce DispatcherRunner 

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

This PR is based on #9806.

In order to better separate concerns which are currently contained in the
Dispatcher, this commit introduces the DispatcherRunner abstraction. The
DispatcherRunner is used by the DispatcherResourceManagerComponent and encapsulates
how the Dispatcher is executed.

## Brief change log

* 940b63c Remove generics from `DispatcherResourceManagerComponent`
* 9ef8fe9 Add `DispatcherRunner`

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Partial {@link DispatcherServices} services container which needs to
+ * be completed before being given to the {@link Dispatcher}.
+ */
+public class PartialDispatcherServices {
```
[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: Actually there is a similar discuss [here](https://github.com/apache/flink/pull/9787).

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: What would be your suggestion for a better name @buptljy? I agree that `DispatcherServices` does not need to strictly inherit from `PartialDispatcherServices`.

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: The intention of the `PartialDispatcherServices` and `DispatcherServices` is to be a container for the dispatcher related services.

[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: With an offline discussion with @buptljy it happens raise that an idea that

```java
public final class DispatcherServicesConstructionHelper { ... }

public /* maybe final */class DispatcherService {
  // ...
  public static DispatcherService from(DispatcherServicesConstructionHelper, JMFactory) { ... }
}
```

for help a bit on naming issue and prevent inheritance leak.

[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: To clarify inherit from `DispatcherService` might be OK for testing but from `PartialDispatcherServices` looks error prone.

[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: Still I am thinking of `DispatcherServices` for sharing services between different Dispatcher leader epoch since we possibly separate concern of leader election and job management from current Dispatcher and let `DispatcherRunner` be the supervisor or `Dispatcher` that start a new instance per leader epoch.

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: Fair enough, I'll remove the inheritance. 

Concerning the renaming, I would like to postpone it until I have merged the remaining PRs in order to fix FLINK-11843 and then do the renaming because otherwise I would end up with quite a bit of merge conflicts.

Concerning the lifecycle management of `DispatcherServices` I think this should be another issue. As I've said, the intention is not to introduce such a lifecycle management but simply to have a service container.

[the remark this answers is not in the document]
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java]: Thanks for sharing your thoughts. You are the coordinator of these batch of pull requests. For concerns of lifecycle management let's move to another thread. I mention here to help express my perspective about the name and inheritance :-) It is simply a service container in this series of patches, definitely.

**Comment by tillrohrmann** [comment]
@flinkbot run travis

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	@Override
+	public T getDispatcher() {
+		return dispatcher;
+	}
+
+	@Override
+	public CompletableFuture<Void> closeAsync() {
+		Collection<CompletableFuture<Void>> terminationFutures = new ArrayList<>(2);
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java]: Can you justify the initial capacity of`2`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+		terminationFutures.add(dispatcher.closeAsync());
+
+		return FutureUtils.completeAll(terminationFutures);
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java]: There is only one future inside `terminationFutures` here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface for a {@link DispatcherRunner} which runs a {@link MiniDispatcher}.
+ */
+public interface MiniDispatcherRunner extends DispatcherRunner {
+
+	@Override
+	MiniDispatcher getDispatcher();
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java]: Why is this needed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunnerFactory.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Factory for the {@link MiniDispatcherRunnerImpl}.
+ */
+public class MiniDispatcherRunnerFactory implements DispatcherRunnerFactory<MiniDispatcherRunnerImpl> {
+
+	private final JobGraphRetriever jobGraphRetriever;
+
+	public MiniDispatcherRunnerFactory(JobGraphRetriever jobGraphRetriever) {
+		this.jobGraphRetriever = jobGraphRetriever;
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunnerFactory.java]: maybe `checkNotNull`

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * The {@link DispatcherRunner} encapsulates how a {@link Dispatcher} is being executed.
+ */
+public interface DispatcherRunner extends AutoCloseableAsync {
+
+	/**
+	 * Get the currently running {@link Dispatcher}.
+	 *
+	 * @return the currently running dispatcher or null if none is running
+	 */
+	Dispatcher getDispatcher();
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java]: `@Nullable` is missing

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java]: I could have. But it's no longer there ;-) I guess this is a good reminder why initial capacities are stupid to set. I'll try to get rid of this habit. Removing now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerImpl.java]: Yes indeed. Will remove it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java]: The `MiniDispatcherRunnerImpl` needs it to get access to the `MiniDispatcher#getJobTerminationFuture()`. This will change in the future (later PR) though.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunnerFactory.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunnerFactory.java]: I've followed the approach that non `@Nullable` annotated parameters must not be null.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java (hunk 6 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java]: I forgot to update the JavaDoc. The return value should always be non-null. Will do it instead of `@Nullable`.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. I've addressed your comments and pushed some fixups.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java (hunk 6 above)
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java]: ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java (hunk 4 above)
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/MiniDispatcherRunner.java]: ok

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. Rebasing and merging this PR.
